src/c/tests/test_node.c

```diff
@@ -0,0 +1,297 @@
+  printf("Queue tests passed!\n");
+}
+
+void test_tree_structure() {
+  int val1 = 1, val2 = 2, val3 = 3;
+  Node *root = node_create(&val1, 1.0);
+  Node *child1 = node_create(&val2, 2.0);
+  Node *child2 = node_create(&val3, 3.0);
+
+  add_child(root, child1);
+  add_child(root, child2);
+
+  assert(node_get_parent(child1) == root);
+  assert(nodeset_size(node_get_children(root)) == 2);
+  assert(is_root(root) == 1);
+  assert(is_leaf(child1) == 1);
+  assert(height(root) == 1);
+  assert(depth(child1) == 1);
+  assert(num_nodes(root) == 3);
+  assert(num_leaves(root) == 2);
+
+  node_destroy(root);
+  node_destroy(child1);
+  node_destroy(child2);
+  printf("Tree structure tests passed!\n");
+}
+
+void test_traversal() {
+  int val1 = 1, val2 = 2;
+  Node *root = node_create(&val1, 1.0);
+  Node *child = node_create(&val2, 2.0);
+  add_child(root, child);
+
+  callback_count = 0;
+  dfs_traverse(root, test_callback);
+  assert(callback_count == 2);
+
+  callback_count = 0;
+  bfs_traverse(root, test_callback);
+  assert(callback_count == 2);
+
+  callback_count = 0;
+  preorder(root, test_callback);
+  assert(callback_count == 2);
+
+  callback_count = 0;
+  postorder(root, test_callback);
+  assert(callback_count == 2);
+
+  node_destroy(root);
+  node_destroy(child);
+  printf("Traversal callback count tests passed!\n");
+}
+
+void test_dfs_output() {
+  Node *root = build_sample_tree();
+  reset_visited();
+
+  dfs_traverse(root, record_callback);
+  int expected[] = {1, 2, 4, 7, 5, 8, 3, 6};
+
+  assert(callback_count == 8);
+  assert(arrays_equal(visited_values, expected, 8));
+  printf("DFS traversal test passed!\n");
+
+  free(root);
+}
+
+void test_bfs_output() {
+  Node *root = build_sample_tree();
+  reset_visited();
+
+  bfs_traverse(root, record_callback);
+  int expected[] = {1, 2, 3, 4, 5, 6, 7, 8};
+
+  assert(callback_count == 8);
+  assert(arrays_equal(visited_values, expected, 8));
+  printf("BFS traversal test passed!\n");
+
+  free(root);
+}
+
+void test_inorder_output() {
+  Node *root = build_sample_bst();
+  reset_visited();
+
+  inorder(root, record_callback);
+  int expected[] = {1, 2, 3, 4, 5, 6, 7};
+
+  assert(callback_count == 7);
+  assert(arrays_equal(visited_values, expected, 7));
+  printf("Inorder traversal test passed!\n");
+
+  free(root);
+}
+
+void test_preorder_output() {
+  Node *root = build_sample_tree();
+  reset_visited();
+
+  preorder(root, record_callback);
+  int expected[] = {1, 2, 4, 7, 5, 8, 3, 6};
+
+  assert(callback_count == 8);
+  assert(arrays_equal(visited_values, expected, 8));
+  printf("Preorder traversal test passed!\n");
+
+  free(root);
+}
+
+void test_postorder_output() {
+  Node *root = build_sample_tree();
+  reset_visited();
+
+  postorder(root, record_callback);
+  int expected[] = {7, 4, 8, 5, 2, 6, 3, 1};
+
+  assert(callback_count == 8);
+  assert(arrays_equal(visited_values, expected, 8));
+  printf("Postorder traversal test passed!\n");
+
+  free(root);
+}
```

Comment:
* there should only be one function that checks all traversal methods.
* drop "callback"
* do assertion based on comparing the expected result with computed result

Reply:
not addressed in the latest commit